ibis/sql/postgres/tests/test_udf.py

```diff
@@ -0,0 +1,173 @@
+"""Test support for already-defined UDFs in Postgres"""
+
+import random
+import string
+
+import pytest
+
+import pandas as pd
+import ibis.expr.datatypes
+import ibis.sql.postgres.udf.api
+from ibis.sql.postgres.client import PostgreSQLClient
+
+
+datatypes = ibis.expr.datatypes
+
+# mark test module as postgresql (for ability to easily exclude,
+# e.g. in conda build tests)
+# (Temporarily adding `postgis` marker so Azure Windows pipeline will exclude
+#     pl/python tests.
+#     TODO: update Windows pipeline to exclude postgres_extensions
+#     TODO: remove postgis marker below once Windows pipeline updated
+pytestmark = [
+    pytest.mark.postgresql,
+    pytest.mark.postgis,
+    pytest.mark.postgres_extensions,
+]
+
+# Database setup (tables and UDFs)
+
+
+test_schema = None
+
+
+def gen_schema_name(basename='test'):
+    """Generate a random alpha string starting with 'test_' to be used as a
+    test schema name"""
+    schema_name = '{}_{}'.format(
+        basename,
+        ''.join([random.choice(string.ascii_lowercase) for i in range(6)])
```

Comment:
Because the testing apparatus often runs in parallel mode, which means parallel processes that don't share globals. No shared globals, no way to do unique incrementing.... (unless we do a Postgres sequence object and base it on the database). Without adding the random strings and checking for existence, when the tests were run in parallel then I was getting contention issues.

Reply:
You can decorate a fixture with `@pytest.fixture(scope='session')` to make the fixture execute exactly once per session. We are mutating state in other backends (notably Impala) so in theory we should be able to do this with postgres as well.

Reply:
This still needs to be addressed. Can you try using `scope='session'`?

Reply:
Just pushed changes with the session-based scope. Also took out the random character-generation method of keeping test schema separate. Commit [e1ee3bbe60](https://github.com/scottcode/ibis/commit/e1ee3bbe601d3d715f6c6a352543c4594e2fa4a1)